extractor/image/image.go

```diff
@@ -0,0 +1,240 @@
+type Image struct {
+	name       string   // e.g. alpine:3.10
+	isFile     bool     // from a tar file
+	transports []string // e.g. "docker://"
+
+	systemContext *imageTypes.SystemContext
+	blobInfoCache imageTypes.BlobInfoCache
+	rawSource     ImageSource
+	src           ImageCloser
+
+	cache cache.Cache
+}
+
+func NewImage(ctx context.Context, image Reference, transports []string, option types.DockerOption,
+	c cache.Cache) (Image, error) {
+	var domain string
+	var auth *imageTypes.DockerAuthConfig
+
+	if !image.IsFile {
+		named, err := reference.ParseNormalizedNamed(image.Name)
+		if err != nil {
+			return Image{}, xerrors.Errorf("invalid image name: %w", err)
+		}
+
+		// add 'latest' tag
+		named = reference.TagNameOnly(named)
+		image.Name = named.String()
+
+		// get a credential for Docker registry
+		domain = reference.Domain(named)
+		auth = GetToken(ctx, domain, option)
+	}
+
+	sys := &imageTypes.SystemContext{
+		// TODO: make OSChoice configurable
+		OSChoice:                          "linux",
+		DockerAuthConfig:                  auth,
+		DockerDisableV1Ping:               option.SkipPing,
+		DockerInsecureSkipTLSVerify:       imageTypes.NewOptionalBool(option.InsecureSkipTLSVerify),
+		OCIInsecureSkipTLSVerify:          option.InsecureSkipTLSVerify,
+		DockerDaemonInsecureSkipTLSVerify: option.InsecureSkipTLSVerify,
+	}
+
+	return Image{
+		name:       image.Name,
+		isFile:     image.IsFile,
+		transports: transports,
+
+		systemContext: sys,
+		blobInfoCache: blobinfocache.DefaultCache(sys),
+
+		cache: c,
+	}, nil
+}
+
+func (img *Image) populateSource() error {
+	if img.rawSource != nil && img.src != nil {
+		return nil
+	}
+	ctx := context.Background()
+	var err error
+	for _, transport := range img.transports {
+		imgName := transport + img.name
+		var ref imageTypes.ImageReference
+		ref, err = alltransports.ParseImageName(imgName)
+		if err != nil {
+			return xerrors.Errorf("failed to parse an image name: %w", err)
+		}
+
+		var rawSource imageTypes.ImageSource
+		rawSource, err = ref.NewImageSource(ctx, img.systemContext)
+		if err != nil {
+			// try next transport
+			continue
+		}
+
+		var src imageTypes.ImageCloser
+		src, err = image.FromSource(ctx, img.systemContext, rawSource)
+		if err != nil {
+			return xerrors.Errorf("failed to initialize: %w", err)
+		}
+
+		img.rawSource = rawSource
+		img.src = src
+		return nil
+	}
+	// return only the last error
+	return err
+}
+
+func (img *Image) LayerInfos() ([]imageTypes.BlobInfo, error) {
+	// ignore the cache error
+	layerInfos, _ := img.getLayerInfosInCache(img.name, img.isFile)
+	if layerInfos != nil {
+		return layerInfos, nil
+	}
+
+	// When it doesn't hit a cache, it fetches the image from Docker Engine or Docker Registry
+	if err := img.populateSource(); err != nil {
+		return nil, xerrors.Errorf("failed population: %w", err)
+	}
+
+	layers := img.src.LayerInfos()
+
+	// ignore the cache error
+	_ = img.storeLayerInfosInCache(img.name, img.isFile, layers)
```

Comment:
Looks like this is the only place where `img.storeLayerInfosInCache` is called. If the error is not important to handle we should remove it from the signature of the method as it is unexported anyways.

Reply:
I think it's up to the caller to ignores the error. `storeLayerInfosInCache` takes only responsibility to store cache and returns an error. But I'm not particular about this. If you prefer not to return an error, I will fix it.

Reply:
Yeah, I'm OK with it. Just thought we could remove some dead code if it will never see the light of the day. Let's keep it for now.